parser_library/test/parsing/string_test.cpp

```diff
@@ -93,106 +94,766 @@ TEST_P(parser_string_fixture, basic)
+
+    if (GetParam().is_consuming)
+    {
+        EXPECT_TRUE(a->diags().empty());
+        EXPECT_EQ(get_var_value<C_t>(a->hlasm_ctx(), "STR"), GetParam().name + "'J");
+    }
+    else
+        EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005" }));
+}
+
+TEST_P(parser_attribute_fixture, text_2_end_apostrophes)
+{
+    std::string input = R"(
+         GBLC &STR
+         MAC )"
+        + GetParam().name + "'J''";
+
+    auto a = analyze(input);
+
+    if (GetParam().is_consuming)
+    {
+        EXPECT_TRUE(a->diags().empty());
+        EXPECT_EQ(get_var_value<C_t>(a->hlasm_ctx(), "STR"), GetParam().name + "'J''");
+    }
+    else
+        EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005" }));
+}
+
+TEST_P(parser_attribute_fixture, text_2_end_apostrophes_remark)
+{
+    std::string input = R"(
+         GBLC &STR
+         MAC )"
+        + GetParam().name + "'J          REMARK''";
+
+    auto a = analyze(input);
+
+    if (GetParam().is_consuming)
+    {
+        EXPECT_TRUE(a->diags().empty());
+        EXPECT_EQ(get_var_value<C_t>(a->hlasm_ctx(), "STR"), GetParam().name + "'J");
+    }
+    else
+        EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005" }));
+}
+
+TEST_P(parser_attribute_fixture, number_variants_2_end_apostrophes)
+{
+    std::string input = R"(
+         GBLC &STR
+         MAC )"
+        + GetParam().name + R"('9''
+         MAC )"
+        + GetParam().name + R"('9           REMARK''
+         MAC )"
+        + GetParam().name + R"('-9''
+         MAC )"
+        + GetParam().name + R"('-9          REMARK''
 )";
-    analyzer a(input);
-    a.analyze();
-    a.collect_diags();
 
-    EXPECT_TRUE(a.diags().empty());
-    EXPECT_EQ(get_var_value<C_t>(a.hlasm_ctx(), "PAR"), "A");
+    auto a = analyze(input);
+
+    EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005", "S0005", "S0005", "S0005" }));
+}
+
+TEST_P(parser_attribute_fixture, list_1_elem_var_instr)
+{
+    std::string input = R"(
+&VAR     SETC 'J'
+         GBLC &STR
+         MAC_LIST_1_ELEM ()"
+        + GetParam().name + "'&VAR)";
+
+    auto a = analyze(input);
+
+    if (GetParam().is_consuming)
+    {
+        EXPECT_TRUE(a->diags().empty());
+        EXPECT_EQ(get_var_value<C_t>(a->hlasm_ctx(), "STR"), GetParam().name + "'J");
+    }
+    else
+        EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005" }));
+}
+
+TEST_P(parser_attribute_fixture, list_1_elem_text)
+{
+    std::string input = R"(
+         GBLC &STR
+         MAC_LIST_1_ELEM ()"
+        + GetParam().name + "'J')";
+
+    auto a = analyze(input);
+
+    if (GetParam().is_consuming)
+        EXPECT_TRUE(matches_message_codes(a->diags(), { "S0005" }));
+    else
+    {
+        EXPECT_TRUE(a->diags().empty());
+        EXPECT_EQ(get_var_value<C_t>(a->hlasm_ctx(), "STR"), GetParam().name + "'J'");
+    }
+}
+
+TEST_P(parser_attribute_fixture, list_1_elem_number)
+{
+    std::string input = R"(
+        GBLC &STR
+        MAC_LIST_1_ELEM ()"
```

Comment:
these things are a bit hard to read - the closing parenthesis is not part of the string - maybe use some sort of placeholder + replace?

Reply:
Refactored